### ISSUE TITLE
Extend file manager to show if and where a file is used

### ISSUE THREAD

**Issue by pgerundt** [contao/contao#1545]
Wouldn't it be nice to have the file manager show information if a file is used somewhere on the website?

I set up a slim C4 extension (tested on 4.8.x and 4.9.x) that adds a new button to the file manager and shows detailed information if and where the file is being used.

You can find the extension here:
[https://github.com/pgerundt/contao-fileusage-bundle](https://github.com/pgerundt/contao-fileusage-bundle)

Any comments would be highly appreciated.

**Comment by m-vo** [comment]
We had this before. Looking through the DCA config isn't sufficient to determine if a file is not being used. If you e.g. use insert tags, references from your layout files, any extension that does add own file fields, [...], this will break.

**Comment by pgerundt** [comment]
Regarding insert tags:
It also works perfectly for insert tags being added into textarea fields. Could be easily modified to check "text" or "inputUnit" fields as well - where you normally do not put insert tags.

Regarding other extensions:
Works perfectly for other extensions.

Regarding "manual" references from layout files:
Yeah, didn't think about that one.

**Comment by m-vo** [comment]
These were just some examples. 

Also:

> Works perfectly for other extensions.

Extension do not need to stick to certain fields, nor even use DCA definitions.

> It also works perfectly for insert tags being added into textarea fields. 

Yeah, and not anywhere else (templates, other fields, custom widgets, ...)

**Comment by pgerundt** [comment]
Well, if the website developer references files in the file manager by url or path (in stylesheets, templates, widgets or whatever), you are right.
If an extension doesn't use DCA definitions, you are right, too.

But the purpose of the extension was to help the website editor (not the developer) in preventing the deletion of files that are currently in use.
And: any (even incomplete) information is far better than having no information at all :)

**Comment by andreheeke** [comment]
I think the core developers are happy about every PR draft they can get. This can be used to further discuss if and how such a feature can be included in one of the next Contao versions. Basically I would like to have such a feature in the core. But in my opinion the file management would have to be reworked anyway.

**Comment by m-vo** [comment]
Don't get me wrong - I'd love to see a working feature like this. I just think a sufficient solution is not possible (nor worth the effort trying to cover and maintain the various use cases).

>And: any (even incomplete) information is far better than having no information at all :)

I disagree on this statement, especially because a feature like this would be used to determine if a seemingly unused file can be deleted.

**Comment by pgerundt** [comment]
> > And: any (even incomplete) information is far better than having no information at all :)
> 
> I disagree on this statement, especially because a feature like this would be used to determine if a seemingly unused file can be deleted.

Well, that depends on the wording of the result message. If an image is found to be used, you could say "This file is (definitely) used in the page/news item/calendar event/... with id X". That would really help the editor a lot (!) - especially when working with a huge pile of files that are traditionally named "test-4.jpg" or "page-header-final-7-800x600.png".

If the file is not used in a fileTree element of Contao (or an insert tag), the message could be "This file doesn't seem to be used by an editor in any Contao element. But before you delete it, make sure blablablabla..."

Advantages of that solution (compared to the current situation):
- Users do get a true positive result when the file is in use.
- The second message is not a false positive result but an indication that the file is not used by an editor.

I don't know. Currently we have 0%, I suggested a solution that covers 80 maybe 90% of all cases and you say "there is no 100% solution" - which is totally correct, but does not get us any further.

**Comment by m-vo** [comment]
Yeah, look at it the other way round: I would have to disable it in any of my setups because it would not work.

But if it's useful for some people in some scenarios it should probably be an extension. :wink: 

The only thing I could think of that would need a centralized solution is providing a registry that 3rd party code could utilize.

**Comment by aschempp** [comment]
> any (even incomplete) information is far better than having no information at all :)

I also disagree, same as @m-vo. Incomplete information is always worse than no information, because you don't know how incomplete it is.

We _could_ think of a *usage statistic* feature, which can search all places whether records are found. But that must only be run _on demand_ because it is slow and inefficient. And third-party developers would need to implement the feature and add their information as well. Certainly a good feature but a lot of effort and probably not easy to implement.

**Comment by fritzmg** [comment]
Additional extensions, that are now available:

- https://extensions.contao.org/?p=sineos%2Ffilemanager-bundle
- https://extensions.contao.org/?p=marcel-mathias-nolte%2Fcontao-filesmanager-fileusage
- https://extensions.contao.org/?p=inspiredminds%2Fcontao-file-usage
